### PR TITLE
[Contributing] Enhance the pull request docs to higher quality

### PR DESCRIPTION
**Pull request by patkar**
| Q | A
| --- | ---
| Fixed tickets | sulu-io/sulu#2073
| Related PRs | sulu-io/sulu#2080, sulu-io/sulu#2085
| License | MIT

#### Description

It goes one in one with the new pull request template in the sulu repo, but it also says clearer that you should rebasing and thus create atomic and logically separate commits. Additionally it says to prefix commits and pull requests already with the affected component for better bisecting and searching on GitHub.

Most of the changes are inspired by the Symfony contributing docs.

#### Purpose

To have high quality pull requests it is not enough to create a pull request template. It must be described in details and especially the checklist needs exlanations

### CODE REVIEW

File: developer/contributing/pull-requests.rst [hunk 1]
```diff
@@ -9,15 +9,37 @@ When creating a pull request:
   example: ``feature/what-my-pr-does``. Note that dashes should be used
   instead of spaces (not underscores).
 * Add your feature/bugfix/enhancement.
+* Do atomic and logically separate commits (use the power of ``git rebase`` to
```
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: s/Do/Create/

File: developer/contributing/pull-requests.rst [hunk 2]
```diff
@@ -9,15 +9,37 @@ When creating a pull request:
...
+  have a clean and logical history).
+* Write good commit messages (see the tip below).
 * Write tests.
 * :doc:`test-your-code`.
 * Add a line in the format of "<type> <pr-number>
   [<affected bundle or component>] <description>" to the ``CHANGELOG.md`` file
   in the root directory.
-* In case you have some changes breaking backwards compability you also have to
+* In case you have some changes breaking backwards compatibility you also have to
```
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: In case you do changes that break backwards…

File: developer/contributing/pull-requests.rst [hunk 3]
```diff
@@ -9,15 +9,37 @@ When creating a pull request:
...
   add a description to the ``UPGRADE.md`` file in the root directory.
-* Use a meaningful name for the pull request.
+* Use a meaningful name for the pull request (see the tip below).
 * Create the pull request as soon as possible.
+* In addition to this "code" pull request, you must also send a pull request to
```
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: In addition to a pull request to the code repository, you should also…

File: developer/contributing/pull-requests.rst [hunk 4]
```diff
@@ -9,15 +9,37 @@ When creating a pull request:
...
+  the `documentation repository`_ to update the documentation when appropriate.
+
+.. tip::
+
+    A good commit message is composed of a summary (the first line),
+    optionally followed by a blank line and a more detailed description. The
+    summary should start with the Component you are working on in square
+    brackets (``[MediaBundle]``, ``[ContentBundle]``, ...). Use a
+    verb (``fixed ...``, ``added ...``, ...) to start the summary and don't
```
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: First letter should be uppercased, i.e. `Fixed ...` etc.

File: developer/contributing/pull-requests.rst [hunk 5]
```diff
@@ -9,15 +9,37 @@ When creating a pull request:
...
 * Create the pull request as soon as possible.
+* In addition to this "code" pull request, you must also send a pull request to
+  the `documentation repository`_ to update the documentation when appropriate.
+
+.. tip::
+
+    A good commit message is composed of a summary (the first line),
+    optionally followed by a blank line and a more detailed description. The
+    summary should start with the Component you are working on in square
+    brackets (``[MediaBundle]``, ``[ContentBundle]``, ...). Use a
```
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: I don't [see this happening](https://github.com/sulu-io/sulu/pulls). Is this rule up to date?

File: developer/contributing/pull-requests.rst [hunk 6]
```diff
@@ -9,15 +9,37 @@ When creating a pull request:
...
+    verb (``fixed ...``, ``added ...``, ...) to start the summary and don't
+    add a period at the end.
+
+.. tip::
+
+    A good pull request title should start with the Component you are working on
+    in square brackets (``[MediaBundle]``, ``[ContentBundle]``, ...). Use a
+    verb (``fixed ...``, ``added ...``, ...) to start the title and don't
```
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: as above

File: developer/contributing/pull-requests.rst [hunk 7]
```diff
@@ -32,67 +54,116 @@ If you are a member of the `Sulu organization`_ you should also:
...
+.. code-block:: text
 
-.. note::
+    | Q                | A
+    | ---------------- | ---
+    | Bug fix?         | [yes|no]
+    | New feature?     | [yes|no]
+    | BC breaks?       | [yes|no]
+    | Deprecations?    | [yes|no]
+    | Tests pass?      | [yes|no]
```
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: Do we actually need this line? Given that CI systems are integrated in GitHub, I think this is superfluous.

File: developer/contributing/pull-requests.rst [hunk 8]
```diff
@@ -32,67 +54,116 @@ If you are a member of the `Sulu organization`_ you should also:
...
+    | Deprecations?    | no
+    | Tests pass?      | yes
+    | Fixed tickets    | fixes #1243, #1443
+    | Related PRs      | none
+    | License          | MIT
+    | Documentation PR | sulu-io/sulu-docs#153
+
+The whole table must be included (do **not** remove lines that you think are
+not relevant). For simple typos, minor changes in the PHPDocs, or changes in
+translation files, use the shorter version of the check-list:
```
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: Isn't this special rule a bit complicated? I'd stick to one rule for all to keep it simple.

Comment (patkar) [inline developer/contributing/pull-requests.rst]: It's from Symfony ;)

File: developer/contributing/pull-requests.rst (hunk 5 above)
Comment (patkar) [inline developer/contributing/pull-requests.rst]: I would like to implement it :grin:

File: developer/contributing/pull-requests.rst (hunk 6 above)
Comment (patkar) [inline developer/contributing/pull-requests.rst]: I would like to implement it :grin:

**Comment by patkar** [comment]
Updated the typos :)

File: developer/contributing/pull-requests.rst (hunk 5 above)
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: Why? This can be automated.

File: developer/contributing/pull-requests.rst (hunk 7 above)
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: For the discussion, see https://github.com/sulu-io/sulu/pull/2080.

**Comment by webmozart** [comment]
See also https://github.com/sulu-io/sulu-docs/pull/181

File: developer/contributing/pull-requests.rst (hunk 5 above)
Comment (patkar) [inline developer/contributing/pull-requests.rst]: How?

**Comment by patkar** [comment]
Removed the tests line, and the to-do point, now its up to if sulu-io/sulu#2080 or sulu-io/sulu#2085 is accepted, thus this docs needs adaptation.

**Comment by chirimoya** [comment]
@patkar would be awesome if you could update. thx

**Comment by patkar** [comment]
@chirimoya @webmozart Docs are updated to reflect the sections from the chosen pull request template.

Any other pain points or considerations? /cc @danrot @wachterjohannes

File: developer/contributing/pull-requests.rst (hunk 5 above)
Comment (patkar) [inline developer/contributing/pull-requests.rst]: @webmozart Should this be enforced or not? It is one step of creating better and atomic commits, as said in sulu-io/sulu#2073

Comment (danrot) [inline developer/contributing/pull-requests.rst]: This is definitely not happening yet. We should probably discuss that altogether /cc @sulu-io/core-developer

File: developer/contributing/pull-requests.rst [hunk 9]
```diff
@@ -9,15 +9,37 @@ When creating a pull request:
...
+    optionally followed by a blank line and a more detailed description. The
+    summary should start with the Component you are working on in square
+    brackets (``[MediaBundle]``, ``[ContentBundle]``, ...). Use a
+    verb (``Fixed ...``, ``Added ...``, ...) to start the summary and don't
+    add a period at the end.
+
+.. tip::
+
+    A good pull request title should start with the Component you are working on
+    in square brackets (``[MediaBundle]``, ``[ContentBundle]``, ...). Use a
```
Comment (danrot) [inline developer/contributing/pull-requests.rst]: Same discussion as above.

File: developer/contributing/pull-requests.rst [hunk 10]
```diff
@@ -9,15 +9,37 @@ When creating a pull request:
...
+    verb (``Fixed ...``, ``Added ...``, ...) to start the title and don't
+    add a period at the end. If your pull request is not finished yet prefix the
+    title with ``[WIP]``.
```
Comment (danrot) [inline developer/contributing/pull-requests.rst]: Prepending ` [WIP]` is also not happening internally because we use [ZenHub](https://www.zenhub.io/) for that. How should we handle this? I don't think there is an easy way to open ZenHub for external users.

File: developer/contributing/pull-requests.rst [hunk 11]
```diff
@@ -32,67 +54,105 @@ If you are a member of the `Sulu organization`_ you should also:
...
+
+* If you answer yes to "Bug fix?", check if the bug is already listed in the
+  Sulu issues and reference it/them in "Fixed tickets". These should be of the
+  form ``fixes #123, fixes #321``.
+* If you answer yes to "New feature?", you must submit a pull request to the
+  documentation and reference it under the "Documentation PR" section.
+* If you answer yes to "BC breaks?", the patch must contain updates to the
+  ``UPGRADE.md`` file.
+* If you answer yes to "Deprecations?", the patch must contain updates to the
+  ``UPGRADE.md`` file.
```
Comment (danrot) [inline developer/contributing/pull-requests.rst]: I would replace `patch` with `pull request` in the previous two bullet points. Using a different term here for the same thing might be confusing.

File: developer/contributing/pull-requests.rst [hunk 12]
```diff
@@ -32,67 +54,105 @@ If you are a member of the `Sulu organization`_ you should also:
...
-    | BC breaks        | none
-    | Documentation PR | https://github.com/sulu-cmf/sulu-docs/pull/14
+.. code-block:: text
+
+    - [ ] Finish the code
+    - [ ] Add tests as they have not been updated yet
+    - [ ] Gather feedback for my changes
+
+As long as you have items in the to-do list, please prefix the pull request
+title with "[WIP]".
```
Comment (danrot) [inline developer/contributing/pull-requests.rst]: Same discussion as above.

File: developer/contributing/pull-requests.rst (hunk 11 above)
Comment (patkar) [inline developer/contributing/pull-requests.rst]: Damn copy pasta ^^

File: developer/contributing/pull-requests.rst (hunk 10 above)
Comment (patkar) [inline developer/contributing/pull-requests.rst]: Oh.. But maybe it must not used internally, but for external contributors. But I understand the discrepancy, maybe remove it.

File: developer/contributing/pull-requests.rst (hunk 11 above)
Comment (danrot) [inline developer/contributing/pull-requests.rst]: Mhm, pasta :spaghetti: :stuck_out_tongue:

File: developer/contributing/pull-requests.rst (hunk 10 above)
Comment (danrot) [inline developer/contributing/pull-requests.rst]: Yeah, that should definitely be consistent... Maybe it would also be an option to remove the review column in ZenHub, and also use the `[WIP]` prefix for internal PRs. Might make the columns less useful, but at least it would be consistent. @sulu-io/core-developer WDYT?

File: developer/contributing/pull-requests.rst (hunk 11 above)
Comment (patkar) [inline developer/contributing/pull-requests.rst]: No more pasta :tired_face:

File: developer/contributing/pull-requests.rst (hunk 5 above)
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: @patkar I think this adds too much bureaucratic burden. We can't expect newcomers to know or follow this rule.

File: developer/contributing/pull-requests.rst (hunk 10 above)
Comment (webmozart) [inline developer/contributing/pull-requests.rst]: I think we should keep the "Review" column. I also don't think we should recommend adding the "[WIP]" prefix. Most of the time, developers of a PR forget to update the title to remove this prefix when a PR is complete. Also, every PR is technically work-in-progress until it is accepted.

My personal stance is that the prefix doesn't add much value.

File: developer/contributing/pull-requests.rst [hunk 13]
```diff
@@ -32,67 +44,102 @@ If you are a member of the `Sulu organization`_ you should also:
...
+    - [ ] Add tests as they have not been updated yet
+    - [ ] Gather feedback for my changes
 
-    | q                | a
-    | ---------------- | ---
-    | Fixed tickets    | fixes #1 , fixes #2
-    | Related PRs      | #3, #4
-    | BC breaks        | none
-    | Documentation PR | https://github.com/sulu-cmf/sulu-docs/pull/14
+Remove the "To Do" section if your pull request is complete and has no tasks.
```
Comment (danrot) [inline developer/contributing/pull-requests.rst]: When thinking of it, I would not remove the Todo section, all the items should just be ticked before merging. The list might also contain information valuable for the reviewer (e.g. which parts of the PR has to be looked very closely at).

File: developer/contributing/pull-requests.rst [hunk 14]
```diff
@@ -32,67 +44,102 @@ If you are a member of the `Sulu organization`_ you should also:
...
+The pull request description helps the code review and it serves as a reference
+when the code is merged.
 
-You can copy and paste the following template:
+Why?
+````
 
-.. code-block:: none
+If your pull request is about adding a new feature or modifying an existing one,
+explain the rationale for the changes. Why you added this feature, which problem
```
Comment (danrot) [inline developer/contributing/pull-requests.rst]: `Why did you add this feature, ...`

**Comment by danrot** [comment]
Sry, have overseen the part with the todo list... But that should be last point from my side.

File: developer/contributing/pull-requests.rst (hunk 13 above)
Comment (patkar) [inline developer/contributing/pull-requests.rst]: Maybe not the best written sentence ^^. Should be "Remove the To Do section if not needed", like code examples.

**Comment by patkar** [comment]
Damn English, updated :)

**Comment by danrot** [comment]
It's fine for me now, if it is also ok for @webmozart, @chirimoya  and @wachterjohannes we can merge that.

**Comment by wachterjohannes** [comment]
+1

**Comment by chirimoya** [comment]
+1

**Comment by webmozart** [comment]
Thanks a lot @patkar :)

**Comment by patkar** [comment]
:neckbeard:
